File: bindings/fortran/tests/standin_test.f90
!> Every call of the Fortran binding (keynub_licdongle module) against a
!! stand-in for the C ABI: bindings/flat/licd_flat.c over
!! bindings/julia/test/stub/licd_stub.c, one imaginary dongle held in memory,
!! linked into the same program. Exit status 0 when every check passed.
!!
!!     cmake -S bindings/fortran/tests -B build-fortran-standin
!!     cmake --build build-fortran-standin
!!     ctest --test-dir build-fortran-standin
program standin_test
    use, intrinsic :: iso_c_binding
    use keynub_licdongle
    implicit none

    character(len=*), parameter :: SERIAL = '04A1B2C3D4E5F6'
    integer(c_int8_t), parameter :: FACTORY_KEY(5) = [int(z'30', c_int8_t), int(z'10', c_int8_t), &
        1_c_int8_t, 2_c_int8_t, 3_c_int8_t]
    integer(c_int8_t), parameter :: REPLACEMENT_KEY(6) = [int(z'30', c_int8_t), int(z'11', c_int8_t), &
        9_c_int8_t, 8_c_int8_t, 7_c_int8_t, 6_c_int8_t]

    integer :: failures = 0
    integer(c_int32_t) :: h, rc, major, minor, patch, count, pa, pb, fa, fb, fc, flags, capacity, free_bytes
    integer(c_int32_t) :: size_out, before, value, scope, i
    character(len=KEYNUB_SERIAL_LEN) :: serial_text
    character(len=KEYNUB_DATE_LEN) :: date_text
    character(len=64) :: name
    logical :: genuine, seen_lic, seen_cfg
    integer(c_int8_t), allocatable :: data(:), blob(:), plain(:)
    integer(c_int8_t) :: root(132), secret(100), nothing(0)

    rc = licdf_version(major, minor, patch)
    call check(rc == LICD_OK .and. major == 9 .and. minor == 8 .and. patch == 7, 'library version')
    call check(keynub_status_text(LICD_E_NO_DEVICE) == 'no device', 'status text')

    rc = licdf_device_count(count)
    call check(rc == LICD_OK .and. count == 1, 'one device')
    call expect(keynub_open('nope'), LICD_E_NO_DEVICE, 'open by unknown serial')

    h = keynub_open()
    call check(h > 0, 'open')
    call expect(keynub_serial(h, serial_text), LICD_OK, 'serial call')
    call check(serial_text == SERIAL, 'serial')
    call expect(licdf_get_info(h, pa, pb, fa, fb, fc, flags, capacity, free_bytes), LICD_OK, 'get_info')
    call check(pa == 1 .and. pb == 0, 'protocol version')
    call check(fa == 2 .and. fb == 3 .and. fc == 4, 'firmware version')
    call check(iand(flags, LICDF_FLAG_SE_READY) /= 0 .and. iand(flags, LICDF_FLAG_PROVISIONED) /= 0 .and. &
               iand(flags, LICDF_FLAG_ISOLATED) /= 0, 'flags set')
    call check(iand(flags, LICDF_FLAG_WATCHDOG_REBOOT) == 0 .and. iand(flags, LICDF_FLAG_WRITEAUTH_ROTATED) == 0, &
               'flags clear')
    call check(capacity == 1048576 .and. free_bytes == 1000000, 'capacity')
    call expect(keynub_verify_genuine(h, genuine, serial_text, date_text), LICD_OK, 'verify_genuine')
    call check(genuine .and. serial_text == SERIAL .and. date_text == '2026-08-15', 'genuine')

    call expect(licdf_set_trust_root(h, [2_c_int8_t, 1_c_int8_t, 0_c_int8_t], 3_c_int32_t), LICD_E_CERT_INVALID, &
                'malformed trust root')
    root = -85_c_int8_t  ! 0xAB
    root(1:4) = [48_c_int8_t, -126_c_int8_t, 1_c_int8_t, 0_c_int8_t]  ! 30 82 01 00
    call expect(licdf_set_trust_root(h, root, 132_c_int32_t), LICD_OK, 'foreign trust root')
    call expect(keynub_verify_genuine(h, genuine), LICD_E_CERT_INVALID, 'verify against a foreign root')
    root(5:) = 1_c_int8_t
    call expect(licdf_set_trust_root(h, root, 132_c_int32_t), LICD_OK, 'right trust root')
    call expect(keynub_verify_genuine(h, genuine), LICD_OK, 'verify after the right root')
    call check(genuine, 'genuine after the right root')

    call expect(licdf_record_count(h, count), LICD_E_SESSION_EXPIRED, 'records without a session')
    call expect(licdf_session_open(h), LICD_OK, 'session_open')
    call expect(keynub_record_write(h, 'lic', bytes('license-blob-0123456789')), LICD_E_AUTH_REQUIRED, &
                'write before the write role')
    call expect(licdf_write_auth(h, [int(z'30', c_int8_t), 0_c_int8_t], 2_c_int32_t), LICD_E_NOT_GENUINE, &
                'write role with a bad key')
    call expect(licdf_write_auth(h, FACTORY_KEY, 5_c_int32_t), LICD_OK, 'write_auth')
    call expect(keynub_record_write(h, 'lic', bytes('license-blob-0123456789')), LICD_OK, 'record_write')
    call expect(keynub_record_read(h, 'lic', data), LICD_OK, 'record_read')
    call check(same(data, bytes('license-blob-0123456789')), 'read back')
    call expect(keynub_record_size(h, 'lic', size_out), LICD_OK, 'record_size')
    call check(size_out == 23, 'record size')
    call expect(keynub_record_write(h, 'cfg', bytes('cfgdata')), LICD_OK, 'second record')
    call expect(licdf_record_count(h, count), LICD_OK, 'record_count')
    call check(count == 2, 'two records')
    seen_lic = .false.
    seen_cfg = .false.
    do i = 0, count - 1
        call expect(keynub_record_name(h, i, name, size_out), LICD_OK, 'record_name')
        if (trim(name) == 'lic' .and. size_out == 23) seen_lic = .true.
        if (trim(name) == 'cfg' .and. size_out == 7) seen_cfg = .true.
    end do
    call check(seen_lic .and. seen_cfg, 'record names and sizes')
    call expect(keynub_record_read(h, 'nope', data), LICD_E_NOT_FOUND, 'read a missing record')
    call check(trim(keynub_last_error(h)) == 'no such record', 'error detail')
    call expect(keynub_record_erase(h, ''), LICD_E_INVALID_ARG, 'erase with an empty name')
    call expect(keynub_record_erase(h, 'cfg'), LICD_OK, 'record_erase')
    call expect(licdf_record_count(h, count), LICD_OK, 'record_count')
    call check(count == 1, 'one record left')
    call expect(keynub_record_write(h, 'empty', nothing), LICD_OK, 'empty record')
    call expect(keynub_record_read(h, 'empty', data), LICD_OK, 'read the empty record')
    call check(size(data) == 0, 'empty record length')

    call expect(licdf_counter_read(h, 0_c_int32_t, before), LICD_OK, 'counter_read')
    call expect(licdf_counter_increment(h, 0_c_int32_t, value), LICD_OK, 'counter_increment')
    call check(value == before + 1, 'increment')
    call expect(licdf_counter_read(h, 1_c_int32_t, value), LICD_OK, 'second counter')
    call check(value == 0, 'counters')
    call expect(licdf_counter_read(h, 7_c_int32_t, value), LICD_E_RANGE, 'counter out of range')

    do i = 1, 100
        secret(i) = int(mod(3 * (i - 1) + 7, 256) - merge(256, 0, mod(3 * (i - 1) + 7, 256) > 127), c_int8_t)
    end do
    do scope = KEYNUB_SCOPE_DEVICE, KEYNUB_SCOPE_DEVELOPER
        call expect(keynub_app_encrypt(h, scope, secret, blob), LICD_OK, 'app_encrypt')
        call check(size(blob) > 100, 'sealed data is longer')
        call check(blob(1) == scope, 'scope byte')
        call expect(keynub_app_decrypt(h, blob, plain), LICD_OK, 'app_decrypt')
        call check(same(plain, secret), 'round trip')
        blob(size(blob)) = ieor(blob(size(blob)), 1_c_int8_t)
        call expect(keynub_app_decrypt(h, blob, plain), LICD_E_TAG_MISMATCH, 'tampered blob')
    end do

    call expect(licdf_record_erase_all(h), LICD_OK, 'record_erase_all')
    call expect(licdf_record_count(h, count), LICD_OK, 'record_count')
    call check(count == 0, 'erase all')

    call expect(licdf_write_auth_rotate(h, REPLACEMENT_KEY, 6_c_int32_t), LICD_OK, 'write_auth_rotate')
    call expect(keynub_record_write(h, 'lic', bytes('still-writable')), LICD_OK, 'still writable')
    call expect(licdf_session_close(h), LICD_OK, 'session_close')
    call expect(licdf_get_info(h, pa, pb, fa, fb, fc, flags, capacity, free_bytes), LICD_OK, 'get_info')
    call check(iand(flags, LICDF_FLAG_WRITEAUTH_ROTATED) /= 0, 'rotated flag')
    call expect(licdf_session_open(h), LICD_OK, 'second session')
    call expect(licdf_write_auth(h, FACTORY_KEY, 5_c_int32_t), LICD_E_NOT_GENUINE, 'factory key after rotation')
    call expect(licdf_write_auth(h, REPLACEMENT_KEY, 6_c_int32_t), LICD_OK, 'new key')
    call expect(keynub_record_write(h, 'lic', bytes('new-key-writes')), LICD_OK, 'write with the new key')
    call expect(keynub_record_read(h, 'lic', data), LICD_OK, 'read with the new key')
    call check(same(data, bytes('new-key-writes')), 'new key content')
    call expect(licdf_session_close(h), LICD_OK, 'session_close')

    call expect(licdf_close(h), LICD_OK, 'close')
    call expect(keynub_serial(h, serial_text), LICD_E_INVALID_ARG, 'serial after close')

    if (failures > 0) then
        print '(I0,A)', failures, ' check(s) failed'
        error stop 1
    end if
    print '(A)', 'keynub_licdongle: every call passed against the ABI stand-in'

contains

    subroutine check(condition, what)
        logical, intent(in) :: condition
        character(len=*), intent(in) :: what
        if (.not. condition) then
            failures = failures + 1
            print '(A,A)', '  FAIL  ', what
        end if
    end subroutine check

    subroutine expect(got, status, what)
        integer(c_int32_t), intent(in) :: got, status
        character(len=*), intent(in) :: what
        if (got /= status) then
            failures = failures + 1
            print '(A,A,A,I0,A,I0)', '  FAIL  ', what, ': ', got, ', expected ', status
        end if
    end subroutine expect

    function bytes(text) result(out)
        character(len=*), intent(in) :: text
        integer(c_int8_t) :: out(len(text))
        out = transfer(text, out)
    end function bytes

    logical function same(a, b)
        integer(c_int8_t), intent(in) :: a(:), b(:)
        same = size(a) == size(b)
        if (same) same = all(a == b)
    end function same

end program standin_test

File: bindings/cpp/tests/standin_test.cpp
// Every call of the C++ wrapper (licdongle.hpp) against a stand-in for the C
// ABI: bindings/julia/test/stub/licd_stub.c, one imaginary dongle held in
// memory, compiled into the same executable. Exit code 0 when every check
// passed.
//
//     cmake -S bindings/cpp/tests -B build-standin && cmake --build build-standin
//     build-standin/standin_test          (from the repository root)
//
// or with the compilers directly (C for the stand-in, C++11 for the test):
//
//     cc -c -DLICD_BUILD_SHARED -Iinclude bindings/julia/test/stub/licd_stub.c -o licd_stub.o
//     c++ -std=c++11 -DLICD_BUILD_SHARED -Iinclude -Ibindings/cpp bindings/cpp/tests/standin_test.cpp licd_stub.o -o standin_test

#include <algorithm>
#include <cstdio>
#include <string>
#include <utility>
#include <vector>

#include "licdongle.hpp"

using keynub::Bytes;
using keynub::Status;

static int failures = 0;

static void check(bool condition, const char *what) {
    if (!condition) {
        failures++;
        std::printf("  FAIL  %s\n", what);
    }
}

template <typename F>
static void fails(Status status, const char *what, F action) {
    try {
        action();
        failures++;
        std::printf("  FAIL  %s: no failure\n", what);
    } catch (const keynub::Error &e) {
        if (e.status() != status) {
            failures++;
            std::printf("  FAIL  %s: %s\n", what, e.what());
        }
    }
}

static Bytes bytes(const std::string &text) { return Bytes(text.begin(), text.end()); }

static const char *kSerial = "04A1B2C3D4E5F6";
static const Bytes kFactoryKey = {0x30, 0x10, 0x01, 0x02, 0x03};
static const Bytes kReplacementKey = {0x30, 0x11, 0x09, 0x08, 0x07, 0x06};

int main() {
    keynub::Version v = keynub::Context::libraryVersion();
    check(v.major == 9 && v.minor == 8 && v.patch == 7, "library version");
    check(v.toString() == "9.8.7", "version text");

    keynub::Context ctx;
    std::vector<keynub::DeviceInfo> found = ctx.enumerate();
    check(found.size() == 1 && found[0].serial == kSerial && found[0].path == "stub:0", "enumerate");
    fails(Status::NoDevice, "open by unknown serial", [&] { ctx.open("nope"); });
    fails(Status::NoDevice, "open by unknown path", [&] { ctx.openPath("stub:9"); });
    try {
        ctx.open("nope");
    } catch (const keynub::DeviceNotFoundError &) {
        check(true, "DeviceNotFoundError");
    } catch (...) {
        check(false, "DeviceNotFoundError");
    }

    keynub::Dongle d = ctx.open();
    check(d.isOpen(), "open");
    check(d.getSerial() == kSerial, "serial");
    keynub::Info i = d.getInfo();
    check(i.protocolMajor == 1 && i.protocolMinor == 0, "protocol version");
    check(i.firmwareMajor == 2 && i.firmwareMinor == 3 && i.firmwarePatch == 4, "firmware version");
    check(i.seReady && i.provisioned && i.isolated, "flags set");
    check(!i.watchdogReboot && !i.writeAuthRotated, "flags clear");
    check(i.dataCapacity == 1024 * 1024 && i.dataFree == 1000000, "capacity");
    keynub::GenuineResult g = d.verifyGenuine();
    check(g.genuine && g.serial == kSerial && g.provisionedDate == "2026-08-15", "genuine");
    check(d.isGenuine(), "isGenuine");

    fails(Status::CertificateInvalid, "malformed trust root", [&] { ctx.setTrustRoot({0x02, 0x01, 0x00}); });
    Bytes root = {0x30, 0x82, 0x01, 0x00};
    root.resize(132, 0xAB);
    ctx.setTrustRoot(root);
    fails(Status::CertificateInvalid, "verify against a foreign root", [&] { d.verifyGenuine(); });
    check(!d.isGenuine(), "isGenuine fails closed");
    std::fill(root.begin() + 4, root.end(), 0x01);
    ctx.setTrustRoot(root);
    check(d.isGenuine(), "isGenuine after the right root");

    {
        keynub::Session s = d.openSession();
        check(s.isOpen(), "session open");
        const Bytes payload = bytes("license-blob-0123456789");
        fails(Status::AuthRequired, "write before the write role", [&] { s.writeRecord("lic", payload); });
        fails(Status::NotGenuine, "write role with a bad key", [&] { s.authorizeWrite({0x30, 0x00}); });
        s.authorizeWrite(kFactoryKey);
        s.writeRecord("lic", payload);
        check(s.readRecord("lic") == payload, "read back");
        s.writeRecord("cfg", bytes("cfgdata"));
        std::vector<keynub::RecordInfo> recs = s.listRecords();
        std::vector<std::string> names;
        for (const keynub::RecordInfo &r : recs) {
            names.push_back(r.name);
        }
        std::sort(names.begin(), names.end());
        check(names == std::vector<std::string>({"cfg", "lic"}), "record names");
        check(std::any_of(recs.begin(), recs.end(),
                          [&](const keynub::RecordInfo &r) { return r.name == "lic" && r.size == payload.size(); }),
              "record size");
        check(s.readRecord("cfg") == bytes("cfgdata"), "second record");
        fails(Status::NotFound, "read a missing record", [&] { s.readRecord("nope"); });
        fails(Status::InvalidArgument, "erase with an empty name", [&] { s.eraseRecord(""); });
        check(s.listRecords().size() == 2, "two records");
        s.eraseRecord("cfg");
        check(s.listRecords().size() == 1 && s.listRecords()[0].name == "lic", "one record left");
        s.writeRecord("empty", Bytes());
        check(s.readRecord("empty").empty(), "empty record");

        // A record bigger than one transfer chunk, with progress reported.
        Bytes big(3000, 0x5A);
        uint32_t lastDone = 0;
        s.writeRecord("big", big, [&](uint32_t done, uint32_t) {
            lastDone = done;
            return true;
        });
        check(lastDone == big.size(), "write progress");
        lastDone = 0;
        check(s.readRecord("big", [&](uint32_t done, uint32_t) {
            lastDone = done;
            return true;
        }) == big, "multi-chunk read");
        check(lastDone == big.size(), "read progress");
        fails(Status::Cancelled, "cancelled transfer", [&] {
            s.readRecord("big", [](uint32_t, uint32_t) { return false; });
        });
        s.eraseRecord("big");

        uint32_t before = s.readCounter(0);
        check(s.incrementCounter(0) == before + 1, "increment");
        check(s.readCounter(0) == before + 1 && s.readCounter(1) == 0, "counters");
        fails(Status::Range, "counter out of range", [&] { s.readCounter(7); });

        Bytes secret(100);
        for (size_t k = 0; k < secret.size(); k++) {
            secret[k] = static_cast<uint8_t>((3 * k + 7) % 256);
        }
        for (keynub::Scope scope : {keynub::Scope::Device, keynub::Scope::Developer}) {
            Bytes blob = s.appEncrypt(scope, secret);
            check(blob.size() > secret.size(), "sealed data is longer");
            check(blob[0] == static_cast<uint8_t>(scope), "scope byte");
            check(s.appDecrypt(blob) == secret, "round trip");
            Bytes tampered = blob;
            tampered.back() ^= 1;
            fails(Status::TagMismatch, "tampered blob", [&] { s.appDecrypt(tampered); });
        }

        s.eraseAllRecords();
        check(s.listRecords().empty(), "erase all");

        s.rotateWriteKey(kReplacementKey);
        s.writeRecord("lic", bytes("still-writable"));
    }
    // The session ended when it left scope.
    check(d.getInfo().writeAuthRotated, "rotated flag");
    {
        keynub::Session s = d.openSession();
        fails(Status::NotGenuine, "factory key after rotation", [&] { s.authorizeWrite(kFactoryKey); });
        s.authorizeWrite(kReplacementKey);
        s.writeRecord("lic", bytes("new-key-writes"));
        check(s.readRecord("lic") == bytes("new-key-writes"), "write with the new key");

        keynub::Session moved = std::move(s);
        check(!s.isOpen() && moved.isOpen(), "session moves");
        moved.close();
        check(!moved.isOpen(), "session close");
        moved.close();
    }

    // A session that outlives its dongle refuses to work instead of touching freed memory.
    keynub::Session orphan = d.openSession();
    d.close();
    check(!d.isOpen(), "dongle close");
    check(!orphan.isOpen(), "session ends with its dongle");
    fails(Status::InvalidArgument, "serial after close", [&] { d.getSerial(); });
    fails(Status::InvalidArgument, "session after its dongle closed", [&] { orphan.readCounter(0); });

    keynub::Dongle again = ctx.open(kSerial);
    keynub::Dongle moved = std::move(again);
    check(moved.isOpen() && !again.isOpen(), "dongle moves");

    if (failures) {
        std::printf("%d check(s) failed\n", failures);
        return 1;
    }
    std::printf("licdongle.hpp: every call passed against the ABI stand-in\n");
    return 0;
}
